{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "=== Downloading necessary NLTK pakages, if not already present\n",
      "=== Loaded\n",
      "Using model ``llama-13b-chat``\n"
     ]
    }
   ],
   "source": [
    "from llamaapi import LlamaAPI\n",
    "import os\n",
    "import nltk\n",
    "\n",
    "from Document import Document\n",
    "\n",
    "print(\"=== Downloading necessary NLTK pakages, if not already present\")\n",
    "\n",
    "d1 = nltk.download(\"punkt\", quiet=True)\n",
    "d2 = nltk.download(\"stopwords\", quiet=True)\n",
    "d3 = nltk.download(\"wordnet\", quiet=True)\n",
    "\n",
    "\n",
    "_MODEL = \"llama-13b-chat\"\n",
    "_CONTEXT_SIZE = 128  # In MB\n",
    "\n",
    "# A bit of obfuscation to make crawlers' life miserable\n",
    "# It's just the a.pi k.ey, written so that it's not a clear string\n",
    "# also, `LlamaAPI' is decoupled from the actual string so as to make it even harder to parse it\n",
    "\n",
    "_a1 = \"AbZ\".split(\"b\")[1]\n",
    "_a2 = \"BSaNbNTlp0o0bILov9Z3U7XmnP4DhwrV24jgq\"\n",
    "_a3 = \"A7kX0SPThAArXd0jNZxQ2WZ\"\n",
    "_call = lambda x: LlamaAPI(x + _a1)\n",
    "\n",
    "llama = _call(f\"LL-{_a3}2l1{_a2}\")\n",
    "\n",
    "\n",
    "print(f\"=== Loaded\\nUsing model ``{_MODEL}``\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_MB_size(path):\n",
    "    file_stats = os.stat(path)\n",
    "    return file_stats.st_size / (1024 * 1024)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\\\\\n\\\\ \\\\\\n\\\\ \\\\\\n\\\\ File:T4\\\\ Bacteriophage\\\\.gifFile:PhageExterior\\\\.svgFile:11\\\\ Hegasy\\\\ Phage\\\\ T4\\\\ Wiki\\\\ E\\\\ CCBYSA\\\\.png\\\\\\n\\\\ A\\\\ bacteriophage\\\\ \\\\(\\\\)\\\\ also\\\\ known\\\\ informally\\\\ as\\\\ a\\\\ phage\\\\ \\\\(\\\\)\\\\ is\\\\ a\\\\ virus\\\\ that\\\\ infects\\\\ and\\\\ replicates\\\\ within\\\\ bacteria\\\\ and\\\\ archaea\\\\.\\\\ The\\\\ term\\\\ was\\\\ derived\\\\ from\\\\ \"bacteria\"\\\\ and\\\\ the\\\\ Greek\\\\ language\\\\\\n\\\\ Bacteriophages\\\\ are\\\\ among\\\\ the\\\\ most\\\\ common\\\\ and\\\\ diverse\\\\ entities\\\\ in\\\\ the\\\\ biosphere\\\\.\\\\ Bacteriophages\\\\ are\\\\ ubiquitous\\\\ viruses\\\\ found\\\\ wherever\\\\ bacteria\\\\ exist\\\\.\\\\ It\\\\ is\\\\ estimated\\\\ there\\\\ are\\\\ more\\\\ than\\\\ 10<sup>31</sup>\\\\ bacteriophages\\\\ on\\\\ the\\\\ planet\\\\ more\\\\ than\\\\ every\\\\ other\\\\ organism\\\\ on\\\\ Earth\\\\ including\\\\ bacteria\\\\ combined\\\\.\\\\ Viruses\\\\ are\\\\ the\\\\ most\\\\ abundant\\\\ biological\\\\ entity\\\\ in\\\\ the\\\\ water\\\\ column\\\\ of\\\\ the\\\\ worlds\\\\ oceans\\\\ and\\\\ the\\\\ second\\\\ largest\\\\ component\\\\ of\\\\ biomass\\\\ after\\\\ prokaryotes\\\\ where\\\\ up\\\\ to\\\\ 9x10<sup>8</sup>\\\\ virus\\\\\\n\\\\ Phages\\\\ have\\\\ been\\\\ used\\\\ since\\\\ the\\\\ late\\\\ 20th\\\\ century\\\\ as\\\\ an\\\\ alternative\\\\ to\\\\ antibiotics\\\\ in\\\\ the\\\\ former\\\\ Soviet\\\\ Union\\\\ and\\\\ Central\\\\ Europe\\\\ as\\\\ well\\\\ as\\\\ in\\\\ France\\\\.\\\\ They\\\\ are\\\\ seen\\\\ as\\\\ a\\\\ possible\\\\ therapy\\\\ against\\\\ multi\\\\-drug\\\\-resistant\\\\ strains\\\\ of\\\\ many\\\\ bacteria\\\\ \\\\(see\\\\ phage\\\\ therapy\\\\)\\\\.\\\\\\n\\\\ \\\\\\n\\\\ Phages\\\\ are\\\\ known\\\\ to\\\\ interact\\\\ with\\\\ the\\\\ immune\\\\ system\\\\ both\\\\ indirectly\\\\ via\\\\ bacterial\\\\ expression\\\\ of\\\\ phage\\\\-encoded\\\\ proteins\\\\ and\\\\ directly\\\\ by\\\\ influencing\\\\ innate\\\\ immunity\\\\ and\\\\ bacterial\\\\ clearance\\\\.\\\\\\n\\\\ \\\\\\n\\\\ \\\\ Classification\\\\ \\\\\\n\\\\ Bacteriophages\\\\ occur\\\\ abundantly\\\\ in\\\\ the\\\\ biosphere\\\\ with\\\\ different\\\\ genomes\\\\ and\\\\ lifestyles\\\\.\\\\ Phages\\\\ are\\\\ classified\\\\ by\\\\ the\\\\ International\\\\ Committee\\\\ on\\\\ Taxonomy\\\\ of\\\\ Viruses\\\\ \\\\(ICTV\\\\)\\\\ according\\\\ to\\\\ morphology\\\\ \\\\(biology\\\\)\\\\\\n\\\\ File:Bacteriophage\\\\ P22\\\\ Casjens\\\\ Lenk\\\\.pngFile:Enterobacteria\\\\ phage\\\\ T2\\\\ transmission\\\\ electron\\\\ micrograph\\\\.jpg!Order\\\\ !!\\\\ Family\\\\ !!\\\\ Morphology\\\\ !!\\\\ Nucleic\\\\ acid\\\\ !!\\\\ Examples\\\\\\n\\\\ \\\\\\n\\\\ It\\\\ has\\\\ been\\\\ suggested\\\\ that\\\\ members\\\\ of\\\\ Picobirnaviridae\\\\ infect\\\\ bacteria\\\\ but\\\\ not\\\\ mammals\\\\.\\\\\\n\\\\ \\\\\\n\\\\ There\\\\ are\\\\ also\\\\ many\\\\ unassigned\\\\ genera\\\\ of\\\\ the\\\\ class\\\\ Leviviricetes:\\\\ Chimpavirus\\\\ Hohglivirus\\\\ Mahrahvirus\\\\ Meihzavirus\\\\ Nicedsevirus\\\\ Sculuvirus\\\\ Skrubnovirus\\\\ Tetipavirus\\\\ and\\\\ Winunavirus\\\\ containing\\\\ linear\\\\ ssRNA\\\\ genomes\\\\ and\\\\ the\\\\ unassigned\\\\ genus\\\\ Lilyvirus\\\\ of\\\\ the\\\\ order\\\\ Caudovirales\\\\ containing\\\\ a\\\\ linear\\\\ dsDNA\\\\ genome\\\\.\\\\\\n\\\\ \\\\\\n\\\\ \\\\ History\\\\ \\\\\\n\\\\ File:Félix\\\\ dHérelle\\\\.jpg\\\\\\n\\\\ In\\\\ 1896\\\\ Ernest\\\\ Hanbury\\\\ Hankin\\\\ reported\\\\ that\\\\ something\\\\ in\\\\ the\\\\ waters\\\\ of\\\\ the\\\\ Ganges\\\\ and\\\\ Yamuna\\\\ rivers\\\\ in\\\\ India\\\\ had\\\\ a\\\\ marked\\\\ antibacterial\\\\ action\\\\ against\\\\ cholera\\\\ and\\\\ it\\\\ could\\\\ pass\\\\ through\\\\ a\\\\ very\\\\ fine\\\\ porcelain\\\\ filter\\\\.\\\\ In\\\\ 1915\\\\ United\\\\ Kingdom\\\\ a\\\\ stage\\\\ in\\\\ the\\\\ Biological\\\\ life\\\\ cycle\\\\ an\\\\ enzyme\\\\ produced\\\\ by\\\\ the\\\\ bacteria\\\\ themselves\\\\ or\\\\\\n\\\\ \\\\ a\\\\ virus\\\\ that\\\\ grew\\\\ on\\\\ and\\\\ destroyed\\\\ the\\\\ bacteria\\\\\\n\\\\ \\\\\\n\\\\ Tworts\\\\ research\\\\ was\\\\ interrupted\\\\ by\\\\ the\\\\ onset\\\\ of\\\\ World\\\\ War\\\\ I\\\\ as\\\\ well\\\\ as\\\\ a\\\\ shortage\\\\ of\\\\ funding\\\\ and\\\\ the\\\\ discoveries\\\\ of\\\\ antibiotics\\\\.\\\\\\n\\\\ \\\\\\n\\\\ Independently\\\\ French\\\\-Canadian\\\\ microbiologist\\\\ Félix\\\\ dHérelle\\\\ working\\\\ at\\\\ the\\\\ Pasteur\\\\ Institute\\\\ in\\\\ Paris\\\\ announced\\\\ on\\\\ 3\\\\ September\\\\ 1917\\\\ that\\\\ he\\\\ had\\\\ discovered\\\\ \"an\\\\ invisible\\\\ antagonistic\\\\ microbe\\\\ of\\\\ the\\\\ dysentery\\\\ bacillus\"\\\\.\\\\ For\\\\ dHérelle\\\\ there\\\\ was\\\\ no\\\\ question\\\\ as\\\\ to\\\\ the\\\\ nature\\\\ of\\\\ his\\\\ discovery:\\\\ \"In\\\\ a\\\\ flash\\\\ I\\\\ had\\\\ understood:\\\\ what\\\\ caused\\\\ my\\\\ clear\\\\ spots\\\\ was\\\\ in\\\\ fact\\\\ an\\\\ invisible\\\\ microbe\\\\.\\\\.\\\\.\\\\ a\\\\ virus\\\\ parasitic\\\\ on\\\\ bacteria\\\\.\"\\\\ DHérelle\\\\ called\\\\ the\\\\ virus\\\\ a\\\\ bacteriophage\\\\ a\\\\ bacteria\\\\-eater\\\\ \\\\(from\\\\ the\\\\ Greek\\\\ \\\\ meaning\\\\ \"to\\\\ devour\"\\\\)\\\\.\\\\ He\\\\ also\\\\ recorded\\\\ a\\\\ dramatic\\\\ account\\\\ of\\\\ a\\\\ man\\\\ suffering\\\\ from\\\\ dysentery\\\\ who\\\\ was\\\\ restored\\\\ to\\\\ good\\\\ health\\\\ by\\\\ the\\\\ bacteriophages\\\\.\\\\ It\\\\ was\\\\ dHérelle\\\\ who\\\\ conducted\\\\ much\\\\ research\\\\ into\\\\ bacteriophages\\\\ and\\\\ introduced\\\\ the\\\\ concept\\\\ of\\\\ phage\\\\ therapy\\\\.\\\\ In\\\\ 1919\\\\ in\\\\ Paris\\\\ France\\\\ dHérelle\\\\ conducted\\\\ the\\\\ first\\\\ clinical\\\\ application\\\\ of\\\\ a\\\\ bacteriophage\\\\ with\\\\ the\\\\ first\\\\ reported\\\\ use\\\\ in\\\\ the\\\\ United\\\\ States\\\\ being\\\\ in\\\\ 1922\\\\.\\\\\\n\\\\ \\\\\\n\\\\ \\\\ Nobel\\\\ prizes\\\\ awarded\\\\ for\\\\ phage\\\\ research\\\\ \\\\\\n\\\\ In\\\\ 1969\\\\ Max\\\\ Delbrück\\\\ Alfred\\\\ Hershey\\\\ and\\\\ Salvador\\\\ Luria\\\\ were\\\\ awarded\\\\ the\\\\ Nobel\\\\ Prize\\\\ in\\\\ Physiology\\\\ or\\\\ Medicine\\\\ for\\\\ their\\\\ discoveries\\\\ of\\\\ the\\\\ replication\\\\ of\\\\ viruses\\\\ and\\\\ their\\\\ genetic\\\\ structure\\\\.\\\\ Specifically\\\\ the\\\\ work\\\\ of\\\\ Hershey\\\\ as\\\\ contributor\\\\ to\\\\ the\\\\ Hershey–Chase\\\\ experiment\\\\ in\\\\ 1952\\\\ provided\\\\ convincing\\\\ evidence\\\\ that\\\\ DNA\\\\ not\\\\ protein\\\\ was\\\\ the\\\\ genetic\\\\ material\\\\ of\\\\ life\\\\.\\\\ Delbrück\\\\ and\\\\ Luria\\\\ carried\\\\ out\\\\ the\\\\ Luria–Delbrück\\\\ experiment\\\\ which\\\\ demonstrated\\\\ statistically\\\\ that\\\\ mutations\\\\ in\\\\ bacteria\\\\ occur\\\\ randomly\\\\ and\\\\ thus\\\\ follow\\\\ Natural\\\\ selection\\\\\\n\\\\ Uses\\\\\\n\\\\ \\\\ Phage\\\\ therapy\\\\ \\\\\\n\\\\ \\\\\\n\\\\ \\\\\\n\\\\ File:George\\\\ Eliava\\\\ 1892–1937\\\\.jpg\\\\\\n\\\\ Phages\\\\ were\\\\ discovered\\\\ to\\\\ be\\\\ antibacterial\\\\ agents\\\\ and\\\\ were\\\\ used\\\\ in\\\\ the\\\\ former\\\\ Soviet\\\\ Republic\\\\ of\\\\ Georgia\\\\ \\\\(country\\\\)\\\\*\\\\ Antibiotics\\\\ were\\\\ discovered\\\\ and\\\\ marketed\\\\ widely\\\\.\\\\ They\\\\ were\\\\ easier\\\\ to\\\\ make\\\\ store\\\\ and\\\\ prescribe\\\\.\\\\\\n\\\\ \\\\*\\\\ Medical\\\\ trials\\\\ of\\\\ phages\\\\ were\\\\ carried\\\\ out\\\\ but\\\\ a\\\\ basic\\\\ lack\\\\ of\\\\ understanding\\\\ of\\\\ phages\\\\ raised\\\\ questions\\\\ about\\\\ the\\\\ validity\\\\ of\\\\ these\\\\ trials\\\\.\\\\\\n\\\\ \\\\*\\\\ Publication\\\\ of\\\\ research\\\\ in\\\\ the\\\\ Soviet\\\\ Union\\\\ was\\\\ mainly\\\\ in\\\\ the\\\\ Russian\\\\ language\\\\\\n\\\\ The\\\\ use\\\\ of\\\\ phages\\\\ has\\\\ continued\\\\ since\\\\ the\\\\ end\\\\ of\\\\ the\\\\ Cold\\\\ War\\\\ in\\\\ Russia\\\\ Georgia\\\\ and\\\\ elsewhere\\\\ in\\\\ Central\\\\ and\\\\ Eastern\\\\ Europe\\\\.\\\\ The\\\\ first\\\\ regulated\\\\ randomized\\\\ double\\\\-blind\\\\ clinical\\\\ trial\\\\ was\\\\ reported\\\\ in\\\\ the\\\\ Journal\\\\ of\\\\ Wound\\\\ Care\\\\ in\\\\ June\\\\ 2009\\\\ which\\\\ evaluated\\\\ the\\\\ safety\\\\ and\\\\ efficacy\\\\ of\\\\ a\\\\ bacteriophage\\\\ cocktail\\\\ to\\\\ treat\\\\ infected\\\\ venous\\\\ ulcers\\\\ of\\\\ the\\\\ leg\\\\ in\\\\ human\\\\ patients\\\\.\\\\ The\\\\ FDA\\\\ approved\\\\ the\\\\ study\\\\ as\\\\ a\\\\ Phase\\\\ I\\\\ clinical\\\\ trial\\\\.\\\\ The\\\\ studys\\\\ results\\\\ demonstrated\\\\ the\\\\ safety\\\\ of\\\\ therapeutic\\\\ application\\\\ of\\\\ bacteriophages\\\\ but\\\\ did\\\\ not\\\\ show\\\\ efficacy\\\\.\\\\ The\\\\ authors\\\\ explained\\\\ that\\\\ the\\\\ use\\\\ of\\\\ certain\\\\ chemicals\\\\ that\\\\ are\\\\ part\\\\ of\\\\ standard\\\\ wound\\\\ care\\\\ \\\\(e\\\\.g\\\\.\\\\ lactoferrin\\\\ or\\\\ silver\\\\)\\\\ may\\\\ have\\\\ interfered\\\\ with\\\\ bacteriophage\\\\ viability\\\\.\\\\ The\\\\ study\\\\ concludes\\\\ that\\\\ bacteriophage\\\\ preparations\\\\ were\\\\ safe\\\\ and\\\\ effective\\\\ for\\\\ treatment\\\\ of\\\\ chronic\\\\ ear\\\\ infections\\\\ in\\\\ humans\\\\.\\\\ Additionally\\\\ there\\\\ have\\\\ been\\\\ numerous\\\\ animal\\\\ and\\\\ other\\\\ experimental\\\\ clinical\\\\ trials\\\\ evaluating\\\\ the\\\\ efficacy\\\\ of\\\\ bacteriophages\\\\ for\\\\ various\\\\ diseases\\\\ such\\\\ as\\\\ infected\\\\ burns\\\\ and\\\\ wounds\\\\ and\\\\ cystic\\\\ fibrosis\\\\-associated\\\\ lung\\\\ infections\\\\ among\\\\ others\\\\.\\\\\\n\\\\ \\\\\\n\\\\ Meanwhile\\\\ bacteriophage\\\\ researchers\\\\ have\\\\ been\\\\ developing\\\\ engineered\\\\ viruses\\\\ to\\\\ overcome\\\\ antimicrobial\\\\ resistance\\\\\\n\\\\ Therapeutic\\\\ efficacy\\\\ of\\\\ a\\\\ phage\\\\ cocktail\\\\ was\\\\ evaluated\\\\ in\\\\ a\\\\ mice\\\\ model\\\\ with\\\\ nasal\\\\ infection\\\\ of\\\\ multidrug\\\\-resistant\\\\ \\\\(MDR\\\\)\\\\ A\\\\.\\\\ baumannii\\\\.\\\\ Mice\\\\ treated\\\\ with\\\\ the\\\\ phage\\\\ cocktail\\\\ showed\\\\ a\\\\ 2\\\\.3\\\\-fold\\\\ higher\\\\ survival\\\\ rate\\\\ compared\\\\ to\\\\ those\\\\ untreated\\\\ at\\\\ seven\\\\ days\\\\ post\\\\-infection\\\\.\\\\ In\\\\ 2017\\\\ a\\\\ patient\\\\ with\\\\ a\\\\ pancreas\\\\ compromised\\\\ by\\\\ MDR\\\\ A\\\\.\\\\ baumannii\\\\ was\\\\ put\\\\ on\\\\ several\\\\ antibiotics;\\\\ despite\\\\ this\\\\ the\\\\ patients\\\\ health\\\\ continued\\\\ to\\\\ deteriorate\\\\ during\\\\ a\\\\ four\\\\-month\\\\ period\\\\.\\\\ Without\\\\ effective\\\\ antibiotics\\\\ the\\\\ patient\\\\ was\\\\ subjected\\\\ to\\\\ phage\\\\ therapy\\\\ using\\\\ a\\\\ phage\\\\ cocktail\\\\ containing\\\\ nine\\\\ different\\\\ phages\\\\ that\\\\ had\\\\ been\\\\ demonstrated\\\\ to\\\\ be\\\\ effective\\\\ against\\\\ MDR\\\\ A\\\\.\\\\ baumannii\\\\.\\\\ Once\\\\ on\\\\ this\\\\ therapy\\\\ the\\\\ patients\\\\ downward\\\\ clinical\\\\ trajectory\\\\ reversed\\\\ and\\\\ returned\\\\ to\\\\ health\\\\.\\\\\\n\\\\ \\\\\\n\\\\ DHerelle\\\\ \"quickly\\\\ learned\\\\ that\\\\ bacteriophages\\\\ are\\\\ found\\\\ wherever\\\\ bacteria\\\\ thrive:\\\\ in\\\\ sewers\\\\ in\\\\ rivers\\\\ that\\\\ catch\\\\ waste\\\\ runoff\\\\ from\\\\ pipes\\\\ and\\\\ in\\\\ the\\\\ stools\\\\ of\\\\ convalescent\\\\ patients\\\\.\"\\\\ This\\\\ includes\\\\ rivers\\\\ traditionally\\\\ thought\\\\ to\\\\ have\\\\ healing\\\\ powers\\\\ including\\\\ Indias\\\\ Ganges\\\\ River\\\\.\\\\\\n\\\\ \\\\\\n\\\\ \\\\ Other\\\\ \\\\\\n\\\\ Food\\\\ industry\\\\ –\\\\ Phages\\\\ have\\\\ increasingly\\\\ been\\\\ used\\\\ to\\\\ safen\\\\ food\\\\ products\\\\ and\\\\ to\\\\ forestall\\\\ spoilage\\\\ bacteria\\\\.\\\\ Since\\\\ 2006\\\\ the\\\\ United\\\\ States\\\\ Food\\\\ and\\\\ Drug\\\\ Administration\\\\ \\\\(FDA\\\\)\\\\ and\\\\ United\\\\ States\\\\ Department\\\\ of\\\\ Agriculture\\\\ \\\\(USDA\\\\)\\\\ have\\\\ approved\\\\ several\\\\ bacteriophage\\\\ products\\\\.\\\\ LMP\\\\-102\\\\ \\\\(Intralytix\\\\)\\\\ was\\\\ approved\\\\ for\\\\ treating\\\\ ready\\\\-to\\\\-eat\\\\ \\\\(RTE\\\\)\\\\ poultry\\\\ and\\\\ meat\\\\ products\\\\.\\\\ In\\\\ that\\\\ same\\\\ year\\\\ the\\\\ FDA\\\\ approved\\\\ LISTEX\\\\ \\\\(developed\\\\ and\\\\ produced\\\\ by\\\\ Micreos\\\\ \\\\(company\\\\)\\\\\\n\\\\ Diagnostics\\\\ –\\\\ In\\\\ 2011\\\\ the\\\\ FDA\\\\ cleared\\\\ the\\\\ first\\\\ bacteriophage\\\\-based\\\\ product\\\\ for\\\\ in\\\\ vitro\\\\ diagnostic\\\\ use\\\\.\\\\ The\\\\ KeyPath\\\\ MRSA/MSSA\\\\ Blood\\\\ Culture\\\\ Test\\\\ uses\\\\ a\\\\ cocktail\\\\ of\\\\ bacteriophage\\\\ to\\\\ detect\\\\ Staphylococcus\\\\ aureus\\\\ in\\\\ positive\\\\ blood\\\\ cultures\\\\ and\\\\ determine\\\\ methicillin\\\\ resistance\\\\ or\\\\ susceptibility\\\\.\\\\ The\\\\ test\\\\ returns\\\\ results\\\\ in\\\\ about\\\\ five\\\\ hours\\\\ compared\\\\ to\\\\ two\\\\ to\\\\ three\\\\ days\\\\ for\\\\ standard\\\\ microbial\\\\ identification\\\\ and\\\\ susceptibility\\\\ test\\\\ methods\\\\.\\\\ It\\\\ was\\\\ the\\\\ first\\\\ accelerated\\\\ antibiotic\\\\-susceptibility\\\\ test\\\\ approved\\\\ by\\\\ the\\\\ FDA\\\\.\\\\\\n\\\\ \\\\\\n\\\\ Counteracting\\\\ bioweapons\\\\ and\\\\ toxins\\\\ –\\\\ Government\\\\ agencies\\\\ in\\\\ the\\\\ West\\\\ have\\\\ for\\\\ several\\\\ years\\\\ been\\\\ looking\\\\ to\\\\ Georgia\\\\ \\\\(country\\\\)\\\\\\n\\\\ The\\\\ sensing\\\\ of\\\\ phage\\\\-triggered\\\\ ion\\\\ cascades\\\\\\n\\\\ Phage\\\\ display\\\\ is\\\\ a\\\\ different\\\\ use\\\\ of\\\\ phages\\\\ involving\\\\ a\\\\ library\\\\ of\\\\ phages\\\\ with\\\\ a\\\\ variable\\\\ peptide\\\\ linked\\\\ to\\\\ a\\\\ surface\\\\ protein\\\\.\\\\ Each\\\\ phage\\\\ genome\\\\ encodes\\\\ the\\\\ variant\\\\ of\\\\ the\\\\ protein\\\\ displayed\\\\ on\\\\ its\\\\ surface\\\\ \\\\(hence\\\\ the\\\\ name\\\\)\\\\ providing\\\\ a\\\\ link\\\\ between\\\\ the\\\\ peptide\\\\ variant\\\\ and\\\\ its\\\\ encoding\\\\ gene\\\\.\\\\ Variant\\\\ phages\\\\ from\\\\ the\\\\ library\\\\ may\\\\ be\\\\ selected\\\\ through\\\\ their\\\\ binding\\\\ affinity\\\\ to\\\\ an\\\\ immobilized\\\\ molecule\\\\ \\\\(e\\\\.g\\\\.\\\\ botulism\\\\ toxin\\\\)\\\\ to\\\\ neutralize\\\\ it\\\\.\\\\ The\\\\ bound\\\\ selected\\\\ phages\\\\ can\\\\ be\\\\ multiplied\\\\ by\\\\ reinfecting\\\\ a\\\\ susceptible\\\\ bacterial\\\\ strain\\\\ thus\\\\ allowing\\\\ them\\\\ to\\\\ retrieve\\\\ the\\\\ peptides\\\\ encoded\\\\ in\\\\ them\\\\ for\\\\ further\\\\ study\\\\.\\\\\\n\\\\ \\\\\\n\\\\ Antimicrobial\\\\ drug\\\\ discovery\\\\ –\\\\ Phage\\\\ proteins\\\\ often\\\\ have\\\\ antimicrobial\\\\ activity\\\\ and\\\\ may\\\\ serve\\\\ as\\\\ leads\\\\ for\\\\ peptidomimetics\\\\ i\\\\.e\\\\.\\\\ drugs\\\\ that\\\\ mimic\\\\ peptides\\\\.\\\\ Phage\\\\-ligand\\\\ technology\\\\ makes\\\\ use\\\\ of\\\\ phage\\\\ proteins\\\\ for\\\\ various\\\\ applications\\\\ such\\\\ as\\\\ binding\\\\ of\\\\ bacteria\\\\ and\\\\ bacterial\\\\ components\\\\ \\\\(e\\\\.g\\\\.\\\\ endotoxin\\\\)\\\\ and\\\\ lysis\\\\ of\\\\ bacteria\\\\.\\\\\\n\\\\ \\\\\\n\\\\ Basic\\\\ research\\\\ –\\\\ Bacteriophages\\\\ are\\\\ important\\\\ model\\\\ organisms\\\\ for\\\\ studying\\\\ principles\\\\ of\\\\ evolution\\\\ and\\\\ ecology\\\\.\\\\\\n\\\\ \\\\\\n\\\\ Detriments\\\\\\n\\\\ Dairy\\\\ industry\\\\\\n\\\\ Bacteriophages\\\\ present\\\\ in\\\\ the\\\\ environment\\\\ can\\\\ cause\\\\ cheese\\\\ to\\\\ not\\\\ ferment\\\\.\\\\ In\\\\ order\\\\ to\\\\ avoid\\\\ this\\\\ mixed\\\\-strain\\\\ starter\\\\ cultures\\\\ and\\\\ culture\\\\ rotation\\\\ regimes\\\\ can\\\\ be\\\\ used\\\\.\\\\ Genetic\\\\ engineering\\\\ of\\\\ culture\\\\ microbes\\\\ –\\\\ especially\\\\ Lactococcus\\\\ lactis\\\\ and\\\\ Streptococcus\\\\ thermophilus\\\\ –\\\\ have\\\\ been\\\\ studied\\\\ for\\\\ genetic\\\\ analysis\\\\ and\\\\ modification\\\\ to\\\\ improve\\\\ phage\\\\ resistance\\\\.\\\\ This\\\\ has\\\\ especially\\\\ focused\\\\ on\\\\ plasmid\\\\ and\\\\ recombinant\\\\ DNA\\\\\\n\\\\ Some\\\\ research\\\\ has\\\\ focused\\\\ on\\\\ the\\\\ potential\\\\ of\\\\ bacteriophages\\\\ as\\\\ antimicrobial\\\\ against\\\\ foodborne\\\\ pathogens\\\\ and\\\\ biofilm\\\\ formation\\\\ within\\\\ the\\\\ dairy\\\\ industry\\\\.\\\\ As\\\\ the\\\\ spread\\\\ of\\\\ antibiotic\\\\ resistance\\\\ is\\\\ a\\\\ main\\\\ concern\\\\ within\\\\ the\\\\ dairy\\\\ industry\\\\ phages\\\\ can\\\\ serve\\\\ as\\\\ a\\\\ promising\\\\ alternative\\\\.\\\\\\n\\\\ \\\\\\n\\\\ \\\\ Replication\\\\ \\\\\\n\\\\ File:Phage\\\\ injection\\\\.pngThe\\\\ life\\\\ cycle\\\\ of\\\\ bacteriophages\\\\ tends\\\\ to\\\\ be\\\\ either\\\\ a\\\\ lytic\\\\ cycle\\\\ or\\\\ a\\\\ lysogenic\\\\ cycle\\\\.\\\\ In\\\\ addition\\\\ some\\\\ phages\\\\ display\\\\ pseudolysogenic\\\\ behaviors\\\\.\\\\\\n\\\\ \\\\\\n\\\\ With\\\\ lytic\\\\ phages\\\\ such\\\\ as\\\\ the\\\\ T4\\\\ phage\\\\ bacterial\\\\ cells\\\\ are\\\\ broken\\\\ open\\\\ \\\\(lysed\\\\)\\\\ and\\\\ destroyed\\\\ after\\\\ immediate\\\\ replication\\\\ of\\\\ the\\\\ virion\\\\.\\\\ As\\\\ soon\\\\ as\\\\ the\\\\ cell\\\\ is\\\\ destroyed\\\\ the\\\\ phage\\\\ progeny\\\\ can\\\\ find\\\\ new\\\\ hosts\\\\ to\\\\ infect\\\\.\\\\ Lytic\\\\ phages\\\\ are\\\\ more\\\\ suitable\\\\ for\\\\ phage\\\\ therapy\\\\.\\\\ Some\\\\ lytic\\\\ phages\\\\ undergo\\\\ a\\\\ phenomenon\\\\ known\\\\ as\\\\ lysis\\\\ inhibition\\\\ where\\\\ completed\\\\ phage\\\\ progeny\\\\ will\\\\ not\\\\ immediately\\\\ lyse\\\\ out\\\\ of\\\\ the\\\\ cell\\\\ if\\\\ extracellular\\\\ phage\\\\ concentrations\\\\ are\\\\ high\\\\.\\\\ This\\\\ mechanism\\\\ is\\\\ not\\\\ identical\\\\ to\\\\ that\\\\ of\\\\ the\\\\ Temperateness\\\\ \\\\(virology\\\\)\\\\\\n\\\\ In\\\\ contrast\\\\ the\\\\ lysogenic\\\\ cycle\\\\ does\\\\ not\\\\ result\\\\ in\\\\ immediate\\\\ lysing\\\\ of\\\\ the\\\\ host\\\\ cell\\\\.\\\\ Those\\\\ phages\\\\ able\\\\ to\\\\ undergo\\\\ lysogeny\\\\ are\\\\ known\\\\ as\\\\ temperate\\\\ phages\\\\.\\\\ Their\\\\ viral\\\\ genome\\\\ will\\\\ integrate\\\\ with\\\\ host\\\\ DNA\\\\ and\\\\ replicate\\\\ along\\\\ with\\\\ it\\\\ relatively\\\\ harmlessly\\\\ or\\\\ may\\\\ even\\\\ become\\\\ established\\\\ as\\\\ a\\\\ plasmid\\\\.\\\\ The\\\\ virus\\\\ remains\\\\ dormant\\\\ until\\\\ host\\\\ conditions\\\\ deteriorate\\\\ perhaps\\\\ due\\\\ to\\\\ depletion\\\\ of\\\\ nutrients\\\\ then\\\\ the\\\\ endogenous\\\\ phages\\\\ \\\\(known\\\\ as\\\\ prophages\\\\)\\\\ become\\\\ active\\\\.\\\\ At\\\\ this\\\\ point\\\\ they\\\\ initiate\\\\ the\\\\ reproductive\\\\ cycle\\\\ resulting\\\\ in\\\\ lysis\\\\ of\\\\ the\\\\ host\\\\ cell\\\\.\\\\ As\\\\ the\\\\ lysogenic\\\\ cycle\\\\ allows\\\\ the\\\\ host\\\\ cell\\\\ to\\\\ continue\\\\ to\\\\ survive\\\\ and\\\\ reproduce\\\\ the\\\\ virus\\\\ is\\\\ replicated\\\\ in\\\\ all\\\\ offspring\\\\ of\\\\ the\\\\ cell\\\\.\\\\ An\\\\ example\\\\ of\\\\ a\\\\ bacteriophage\\\\ known\\\\ to\\\\ follow\\\\ the\\\\ lysogenic\\\\ cycle\\\\ and\\\\ the\\\\ lytic\\\\ cycle\\\\ is\\\\ the\\\\ phage\\\\ lambda\\\\ of\\\\ E\\\\.\\\\ coli\\\\.\\\\\\n\\\\ \\\\\\n\\\\ Sometimes\\\\ prophages\\\\ may\\\\ provide\\\\ benefits\\\\ to\\\\ the\\\\ host\\\\ bacterium\\\\ while\\\\ they\\\\ are\\\\ dormant\\\\ by\\\\ adding\\\\ new\\\\ functions\\\\ to\\\\ the\\\\ bacterial\\\\ genome\\\\ in\\\\ a\\\\ phenomenon\\\\ called\\\\ lysogenic\\\\ conversion\\\\.\\\\ Examples\\\\ are\\\\ the\\\\ conversion\\\\ of\\\\ harmless\\\\ strains\\\\ of\\\\ Corynebacterium\\\\ diphtheriae\\\\ or\\\\ Vibrio\\\\ cholerae\\\\ by\\\\ bacteriophages\\\\ to\\\\ highly\\\\ virulent\\\\ ones\\\\ that\\\\ cause\\\\ diphtheria\\\\ or\\\\ cholera\\\\ respectively\\\\.\\\\ Strategies\\\\ to\\\\ combat\\\\ certain\\\\ bacterial\\\\ infections\\\\ by\\\\ targeting\\\\ these\\\\ toxin\\\\-encoding\\\\ prophages\\\\ have\\\\ been\\\\ proposed\\\\.\\\\\\n\\\\ \\\\\\n\\\\ \\\\ Attachment\\\\ and\\\\ penetration\\\\ \\\\\\n\\\\ \\\\\\n\\\\ File:Phage\\\\.jpg<!\\\\-\\\\-\\\\ Deleted\\\\ image\\\\ removed:\\\\ File:Trialphage\\\\.jpgBacterial\\\\ cells\\\\ are\\\\ protected\\\\ by\\\\ a\\\\ cell\\\\ wall\\\\ of\\\\ polysaccharides\\\\ which\\\\ are\\\\ important\\\\ virulence\\\\ factors\\\\ protecting\\\\ bacterial\\\\ cells\\\\ against\\\\ both\\\\ immune\\\\ host\\\\ defenses\\\\ and\\\\ antibiotics\\\\.\\\\ \\\\\\n\\\\ Host\\\\ growth\\\\ conditions\\\\ also\\\\ influence\\\\ the\\\\ ability\\\\ of\\\\ the\\\\ phage\\\\ to\\\\ attach\\\\ and\\\\ invade\\\\ them\\\\.\\\\ As\\\\ phage\\\\ virions\\\\ do\\\\ not\\\\ move\\\\ independently\\\\ they\\\\ must\\\\ rely\\\\ on\\\\ random\\\\ encounters\\\\ with\\\\ the\\\\ correct\\\\ receptors\\\\ when\\\\ in\\\\ solution\\\\ such\\\\ as\\\\ blood\\\\ lymphatic\\\\ circulation\\\\ irrigation\\\\ soil\\\\ water\\\\ etc\\\\.\\\\\\n\\\\ \\\\\\n\\\\ Myovirus\\\\ bacteriophages\\\\ use\\\\ a\\\\ hypodermic\\\\ syringe\\\\-like\\\\ motion\\\\ to\\\\ inject\\\\ their\\\\ genetic\\\\ material\\\\ into\\\\ the\\\\ cell\\\\.\\\\ After\\\\ contacting\\\\ the\\\\ appropriate\\\\ receptor\\\\ the\\\\ tail\\\\ fibers\\\\ flex\\\\ to\\\\ bring\\\\ the\\\\ base\\\\ plate\\\\ closer\\\\ to\\\\ the\\\\ surface\\\\ of\\\\ the\\\\ cell\\\\.\\\\ This\\\\ is\\\\ known\\\\ as\\\\ reversible\\\\ binding\\\\.\\\\ Once\\\\ attached\\\\ completely\\\\ irreversible\\\\ binding\\\\ is\\\\ initiated\\\\ and\\\\ the\\\\ tail\\\\ contracts\\\\ possibly\\\\ with\\\\ the\\\\ help\\\\ of\\\\ Adenosine\\\\ triphosphate\\\\\\n\\\\ \\\\ Synthesis\\\\ of\\\\ proteins\\\\ and\\\\ nucleic\\\\ acid\\\\ \\\\\\n\\\\ Within\\\\ minutes\\\\ bacterial\\\\ ribosomes\\\\ start\\\\ translating\\\\ viral\\\\ mRNA\\\\ into\\\\ protein\\\\.\\\\ For\\\\ RNA\\\\-based\\\\ phages\\\\ RNA\\\\ replicase\\\\ is\\\\ synthesized\\\\ early\\\\ in\\\\ the\\\\ process\\\\.\\\\ Proteins\\\\ modify\\\\ the\\\\ bacterial\\\\ RNA\\\\ polymerase\\\\ so\\\\ it\\\\ preferentially\\\\ transcribes\\\\ viral\\\\ mRNA\\\\.\\\\ The\\\\ hosts\\\\ normal\\\\ synthesis\\\\ of\\\\ proteins\\\\ and\\\\ nucleic\\\\ acids\\\\ is\\\\ disrupted\\\\ and\\\\ it\\\\ is\\\\ forced\\\\ to\\\\ manufacture\\\\ viral\\\\ products\\\\ instead\\\\.\\\\ These\\\\ products\\\\ go\\\\ on\\\\ to\\\\ become\\\\ part\\\\ of\\\\ new\\\\ virions\\\\ within\\\\ the\\\\ cell\\\\ helper\\\\ proteins\\\\ that\\\\ contribute\\\\ to\\\\ the\\\\ assemblage\\\\ of\\\\ new\\\\ virions\\\\ or\\\\ proteins\\\\ involved\\\\ in\\\\ cell\\\\ lysis\\\\.\\\\ In\\\\ 1972\\\\ Walter\\\\ Fiers\\\\ \\\\(University\\\\ of\\\\ Ghent\\\\ Belgium\\\\)\\\\ was\\\\ the\\\\ first\\\\ to\\\\ establish\\\\ the\\\\ complete\\\\ nucleotide\\\\ sequence\\\\ of\\\\ a\\\\ gene\\\\ and\\\\ in\\\\ 1976\\\\ of\\\\ the\\\\ viral\\\\ genome\\\\ of\\\\ bacteriophage\\\\ MS2\\\\.\\\\ Some\\\\ DNAssDNA\\\\ vs\\\\.\\\\ dsDNA\\\\\\n\\\\ \\\\ Virion\\\\ assembly\\\\ \\\\\\n\\\\ In\\\\ the\\\\ case\\\\ of\\\\ the\\\\ T4\\\\ phage\\\\ the\\\\ construction\\\\ of\\\\ new\\\\ virus\\\\ particles\\\\ involves\\\\ the\\\\ assistance\\\\ of\\\\ helper\\\\ proteins\\\\ that\\\\ act\\\\ catalytically\\\\ during\\\\ phage\\\\ morphogenesis\\\\.\\\\ The\\\\ base\\\\ plates\\\\ are\\\\ assembled\\\\ first\\\\ with\\\\ the\\\\ tails\\\\ being\\\\ built\\\\ upon\\\\ them\\\\ afterward\\\\.\\\\ The\\\\ head\\\\ capsids\\\\ constructed\\\\ separately\\\\ will\\\\ spontaneously\\\\ assemble\\\\ with\\\\ the\\\\ tails\\\\.\\\\ \\\\ During\\\\ assembly\\\\ of\\\\ the\\\\ Escherichia\\\\ virus\\\\ T4\\\\\\n\\\\ \\\\ Release\\\\ of\\\\ virions\\\\ \\\\\\n\\\\ Phages\\\\ may\\\\ be\\\\ released\\\\ via\\\\ cell\\\\ lysis\\\\ by\\\\ extrusion\\\\ or\\\\ in\\\\ a\\\\ few\\\\ cases\\\\ by\\\\ budding\\\\.\\\\ Lysis\\\\ by\\\\ tailed\\\\ phages\\\\ is\\\\ achieved\\\\ by\\\\ an\\\\ enzyme\\\\ called\\\\ endolysin\\\\ which\\\\ attacks\\\\ and\\\\ breaks\\\\ down\\\\ the\\\\ cell\\\\ wall\\\\ peptidoglycan\\\\.\\\\ An\\\\ altogether\\\\ different\\\\ phage\\\\ type\\\\ the\\\\ filamentous\\\\ phage\\\\ makes\\\\ the\\\\ host\\\\ cell\\\\ continually\\\\ secrete\\\\ new\\\\ virus\\\\ particles\\\\.\\\\ Released\\\\ virions\\\\ are\\\\ described\\\\ as\\\\ free\\\\ and\\\\ unless\\\\ defective\\\\ are\\\\ capable\\\\ of\\\\ infecting\\\\ a\\\\ new\\\\ bacterium\\\\.\\\\ Budding\\\\ is\\\\ associated\\\\ with\\\\ certain\\\\ Mycoplasma\\\\ phages\\\\.\\\\ In\\\\ contrast\\\\ to\\\\ virion\\\\ release\\\\ phages\\\\ displaying\\\\ a\\\\ lysogenic\\\\ cycle\\\\ do\\\\ not\\\\ kill\\\\ the\\\\ host\\\\ and\\\\ instead\\\\ become\\\\ long\\\\-term\\\\ residents\\\\ as\\\\ prophages\\\\.\\\\\\n\\\\ \\\\\\n\\\\ \\\\ Communication\\\\ \\\\\\n\\\\ Research\\\\ in\\\\ 2017\\\\ revealed\\\\ that\\\\ the\\\\ bacteriophage\\\\ Φ3T\\\\ makes\\\\ a\\\\ short\\\\ viral\\\\ protein\\\\ that\\\\ signals\\\\ other\\\\ bacteriophages\\\\ to\\\\ lie\\\\ dormant\\\\ instead\\\\ of\\\\ killing\\\\ the\\\\ host\\\\ bacterium\\\\.\\\\ Arbitrium\\\\ is\\\\ the\\\\ name\\\\ given\\\\ to\\\\ this\\\\ protein\\\\ by\\\\ the\\\\ researchers\\\\ who\\\\ discovered\\\\ it\\\\.\\\\\\n\\\\ \\\\\\n\\\\ \\\\ Genome\\\\ structure\\\\ \\\\\\n\\\\ Given\\\\ the\\\\ millions\\\\ of\\\\ different\\\\ phages\\\\ in\\\\ the\\\\ environment\\\\ phage\\\\ genomes\\\\ come\\\\ in\\\\ a\\\\ variety\\\\ of\\\\ forms\\\\ and\\\\ sizes\\\\.\\\\ RNA\\\\ phages\\\\ such\\\\ as\\\\ Bacteriophage\\\\ MS2Some\\\\ marine\\\\ roseobacter\\\\ phages\\\\ contain\\\\ deoxyuridine\\\\ \\\\(dU\\\\)\\\\ instead\\\\ of\\\\ Thymidine\\\\\\n\\\\ \\\\ Systems\\\\ biology\\\\ \\\\\\n\\\\ The\\\\ field\\\\ of\\\\ systems\\\\ biology\\\\ investigates\\\\ the\\\\ complex\\\\ Biological\\\\ network\\\\\\n\\\\ For\\\\ instance\\\\ infection\\\\ of\\\\ Pseudomonas\\\\ aeruginosa\\\\ by\\\\ the\\\\ temperate\\\\ phage\\\\ PaP3\\\\ changed\\\\ the\\\\ expression\\\\ of\\\\ 38%\\\\ \\\\(2160/5633\\\\)\\\\ of\\\\ its\\\\ hosts\\\\ genes\\\\.\\\\ Many\\\\ of\\\\ these\\\\ effects\\\\ are\\\\ probably\\\\ indirect\\\\ hence\\\\ the\\\\ challenge\\\\ becomes\\\\ to\\\\ identify\\\\ the\\\\ direct\\\\ interactions\\\\ among\\\\ bacteria\\\\ and\\\\ phage\\\\.\\\\\\n\\\\ \\\\\\n\\\\ Several\\\\ attempts\\\\ have\\\\ been\\\\ made\\\\ to\\\\ map\\\\ protein–protein\\\\ interactions\\\\ among\\\\ phage\\\\ and\\\\ their\\\\ host\\\\.\\\\ For\\\\ instance\\\\ bacteriophage\\\\ lambda\\\\ was\\\\ found\\\\ to\\\\ interact\\\\ with\\\\ its\\\\ host\\\\ Escherichia\\\\ coli\\\\\\n\\\\ \\\\ Host\\\\ resistance\\\\ \\\\\\n\\\\ Bacteriophages\\\\ are\\\\ a\\\\ major\\\\ threat\\\\ to\\\\ bacteria\\\\ and\\\\ prokaryotes\\\\ have\\\\ evolved\\\\ numerous\\\\ mechanisms\\\\ to\\\\ block\\\\ infection\\\\ or\\\\ to\\\\ block\\\\ the\\\\ replication\\\\ of\\\\ bacteriophages\\\\ within\\\\ host\\\\ cells\\\\.\\\\ The\\\\ CRISPR\\\\\\n\\\\ \\\\ Bacteriophage–host\\\\ symbiosis\\\\ \\\\\\n\\\\ Temperate\\\\ phages\\\\ are\\\\ bacteriophages\\\\ that\\\\ integrate\\\\ their\\\\ genetic\\\\ material\\\\ into\\\\ the\\\\ host\\\\ as\\\\ extrachromosomal\\\\ episomes\\\\ or\\\\ as\\\\ a\\\\ prophage\\\\ during\\\\ a\\\\ lysogenic\\\\ cycle\\\\.\\\\ Some\\\\ temperate\\\\ phages\\\\ can\\\\ confer\\\\ fitness\\\\ advantages\\\\ to\\\\ their\\\\ host\\\\ in\\\\ numerous\\\\ ways\\\\ including\\\\ giving\\\\ antibiotic\\\\ resistance\\\\ through\\\\ the\\\\ transfer\\\\ or\\\\ introduction\\\\ of\\\\ antibiotic\\\\ resistance\\\\ genes\\\\ \\\\(ARGs\\\\)\\\\ protecting\\\\ hosts\\\\ from\\\\ phagocytosis\\\\ protecting\\\\ hosts\\\\ from\\\\ secondary\\\\ infection\\\\ through\\\\ superinfection\\\\ exclusion\\\\ enhancing\\\\ host\\\\ pathogenicity\\\\ or\\\\ enhancing\\\\ bacterial\\\\ metabolism\\\\ or\\\\ growth\\\\.\\\\ Bacteriophage–host\\\\ symbiosis\\\\ may\\\\ benefit\\\\ bacteria\\\\ by\\\\ providing\\\\ selective\\\\ advantages\\\\ while\\\\ passively\\\\ replicating\\\\ the\\\\ phage\\\\ genome\\\\.\\\\\\n\\\\ \\\\\\n\\\\ \\\\ In\\\\ the\\\\ environment\\\\ \\\\\\n\\\\ \\\\\\n\\\\ Metagenomics\\\\ has\\\\ allowed\\\\ the\\\\ in\\\\-water\\\\ detection\\\\ of\\\\ bacteriophages\\\\ that\\\\ was\\\\ not\\\\ possible\\\\ previously\\\\.\\\\\\n\\\\ \\\\\\n\\\\ Also\\\\ bacteriophages\\\\ have\\\\ been\\\\ used\\\\ in\\\\ hydrology\\\\\\n\\\\ Bacteriophages\\\\ are\\\\ thought\\\\ to\\\\ contribute\\\\ extensively\\\\ to\\\\ horizontal\\\\ gene\\\\ transfer\\\\ in\\\\ natural\\\\ environments\\\\ principally\\\\ via\\\\ transduction\\\\ \\\\(genetics\\\\)\\\\\\n\\\\ \\\\ In\\\\ humans\\\\ \\\\\\n\\\\ Although\\\\ phages\\\\ do\\\\ not\\\\ infect\\\\ humans\\\\ there\\\\ are\\\\ countless\\\\ phage\\\\ particles\\\\ in\\\\ the\\\\ human\\\\ body\\\\ given\\\\ our\\\\ extensive\\\\ microbiome\\\\.\\\\ Our\\\\ phage\\\\ population\\\\ has\\\\ been\\\\ called\\\\ the\\\\ human\\\\ phageome\\\\ including\\\\ the\\\\ \"healthy\\\\ gut\\\\ phageome\"\\\\ \\\\(HGP\\\\)\\\\ and\\\\ the\\\\ \"diseased\\\\ human\\\\ phageome\"\\\\ \\\\(DHP\\\\)\\\\.\\\\ The\\\\ active\\\\ phageome\\\\ of\\\\ a\\\\ healthy\\\\ human\\\\ \\\\(i\\\\.e\\\\.\\\\ actively\\\\ replicating\\\\ as\\\\ opposed\\\\ to\\\\ nonreplicating\\\\ integrated\\\\ prophage\\\\)\\\\ has\\\\ been\\\\ estimated\\\\ to\\\\ comprise\\\\ dozens\\\\ to\\\\ thousands\\\\ of\\\\ different\\\\ viruses\\\\.\\\\\\n\\\\ There\\\\ is\\\\ evidence\\\\ that\\\\ bacteriophages\\\\ and\\\\ bacteria\\\\ interact\\\\ in\\\\ the\\\\ \\\\ Gut\\\\ microbiota\\\\\\n\\\\ Preliminary\\\\ studies\\\\ have\\\\ indicated\\\\ that\\\\ common\\\\ bacteriophages\\\\ are\\\\ found\\\\ in\\\\ 62%\\\\ of\\\\ healthy\\\\ individuals\\\\ on\\\\ average\\\\ while\\\\ their\\\\ prevalence\\\\ was\\\\ reduced\\\\ by\\\\ 42%\\\\ and\\\\ 54%\\\\ on\\\\ average\\\\ in\\\\ patients\\\\ with\\\\ ulcerative\\\\ colitis\\\\ \\\\(UC\\\\)\\\\ and\\\\ Crohns\\\\ disease\\\\ \\\\(CD\\\\)\\\\.\\\\ Abundance\\\\ of\\\\ phages\\\\ may\\\\ also\\\\ decline\\\\ in\\\\ the\\\\ elderly\\\\.\\\\\\n\\\\ \\\\\\n\\\\ The\\\\ most\\\\ common\\\\ phages\\\\ in\\\\ the\\\\ human\\\\ intestine\\\\ found\\\\ worldwide\\\\ are\\\\ crAssphages\\\\.\\\\ CrAssphages\\\\ are\\\\ transmitted\\\\ from\\\\ mother\\\\ to\\\\ child\\\\ soon\\\\ after\\\\ birth\\\\ and\\\\ there\\\\ is\\\\ some\\\\ evidence\\\\ suggesting\\\\ that\\\\ they\\\\ may\\\\ be\\\\ transmitted\\\\ locally\\\\.\\\\ Each\\\\ person\\\\ develops\\\\ their\\\\ own\\\\ unique\\\\ crAssphage\\\\ clusters\\\\.\\\\ CrAss\\\\-like\\\\ phages\\\\ also\\\\ may\\\\ be\\\\ present\\\\ in\\\\ primates\\\\ besides\\\\ humans\\\\.\\\\\\n\\\\ \\\\\\n\\\\ \\\\ Commonly\\\\ studied\\\\ bacteriophage\\\\ \\\\\\n\\\\ Among\\\\ the\\\\ countless\\\\ phage\\\\ only\\\\ a\\\\ few\\\\ have\\\\ been\\\\ studied\\\\ in\\\\ detail\\\\ including\\\\ some\\\\ historically\\\\ important\\\\ phage\\\\ that\\\\ were\\\\ discovered\\\\ in\\\\ the\\\\ early\\\\ days\\\\ of\\\\ microbial\\\\ genetics\\\\.\\\\ These\\\\ especially\\\\ the\\\\ T\\\\-phage\\\\ helped\\\\ to\\\\ discover\\\\ important\\\\ principles\\\\ of\\\\ gene\\\\ structure\\\\ and\\\\ function\\\\.\\\\\\n\\\\ \\\\\\n\\\\ \\\\*\\\\ 186\\\\ phage\\\\\\n\\\\ \\\\*\\\\ Lambda\\\\ phage\\\\*\\\\ Pseudomonas\\\\ phage\\\\ Φ6\\\\*\\\\ Φ29\\\\ phage\\\\\\n\\\\ \\\\*\\\\ Phi\\\\ X\\\\ 174\\\\*\\\\ Bacteriophage\\\\ φCb5\\\\\\n\\\\ \\\\*\\\\ G4\\\\ phage\\\\\\n\\\\ \\\\*\\\\ M13\\\\ phage\\\\\\n\\\\ \\\\*\\\\ MS2\\\\ phage\\\\ \\\\(23–28\\\\ nanometre\\\\*\\\\ N4\\\\ phage\\\\\\n\\\\ \\\\*\\\\ P1\\\\ phage\\\\\\n\\\\ \\\\*\\\\ Enterobacteria\\\\ phage\\\\ P2\\\\*\\\\ P4\\\\ phage\\\\\\n\\\\ \\\\*\\\\ R17\\\\ phage\\\\\\n\\\\ \\\\*\\\\ T2\\\\ phage\\\\\\n\\\\ \\\\*\\\\ T4\\\\ phage\\\\ \\\\(169\\\\ Base\\\\ pair\\\\*\\\\ T7\\\\ phage\\\\\\n\\\\ \\\\*\\\\ T12\\\\ phage\\\\\\n\\\\ \\\\\\n\\\\ \\\\\\n\\\\ \\\\ See\\\\ also\\\\ \\\\\\n\\\\ \\\\\\n\\\\ \\\\*\\\\ Bacterivore\\\\\\n\\\\ \\\\*\\\\ CrAssphage\\\\\\n\\\\ \\\\*\\\\ CRISPR\\\\\\n\\\\ \\\\*\\\\ DNA\\\\ viruses\\\\\\n\\\\ \\\\*\\\\ Macrophage\\\\\\n\\\\ \\\\*\\\\ Phage\\\\ ecology\\\\\\n\\\\ \\\\*\\\\ Phage\\\\ monographs\\\\ \\\\(a\\\\ comprehensive\\\\ listing\\\\ of\\\\ phage\\\\ and\\\\ phage\\\\-associated\\\\ monographs\\\\ 1921–present\\\\)\\\\\\n\\\\ \\\\*\\\\ Phagemid\\\\\\n\\\\ \\\\*\\\\ Polyphage\\\\\\n\\\\ \\\\*\\\\ RNA\\\\ viruses\\\\\\n\\\\ \\\\*\\\\ Transduction\\\\ \\\\(genetics\\\\)\\\\*\\\\ Viriome\\\\\\n\\\\ \\\\*\\\\ Virophage\\\\ viruses\\\\ that\\\\ infect\\\\ other\\\\ viruses\\\\\\n\\\\ \\\\\\n\\\\ \\\\ References\\\\ \\\\\\n\\\\ \\\\\\n\\\\ \\\\\\n\\\\ \\\\ Bibliography\\\\ \\\\\\n\\\\ \\\\\\n\\\\ \\\\*\\\\ \\\\\\n\\\\ \\\\*\\\\ \\\\\\n\\\\ \\\\*\\\\ \\\\\\n\\\\ \\\\\\n\\\\ \\\\\\n\\\\ \\\\ External\\\\ links\\\\ \\\\\\n\\\\ \\\\\\n\\\\ \\\\\\n\\\\ \\\\*\\\\ \\\\\\n\\\\ \\\\*\\\\ \\\\\\n\\\\ \\\\*\\\\ \\\\\\n\\\\ \\\\*\\\\ \\\\\\n\\\\ \\\\*\\\\ \\\\\\n\\\\ \\\\*\\\\ \\\\\\n\\\\ \\\\*\\\\ \\\\\\n\\\\ \\\\\\n\\\\ \\\\\\n\\\\ \\\\\\n\\\\ \\\\\\n\\\\ \\\\\\n'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path = \"./test.txt\"\n",
    "\n",
    "if get_MB_size(path) > _CONTEXT_SIZE:\n",
    "    print(\"Over the limit\")\n",
    "document = Document(path).make_bow()\n",
    "# document.counts()\n",
    "text = document.text(escape=True)\n",
    "text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "ename": "JSONDecodeError",
     "evalue": "Expecting value: line 1 column 1 (char 0)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mJSONDecodeError\u001b[0m                           Traceback (most recent call last)",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/requests/models.py:971\u001b[0m, in \u001b[0;36mResponse.json\u001b[0;34m(self, **kwargs)\u001b[0m\n\u001b[1;32m    970\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 971\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m complexjson\u001b[38;5;241m.\u001b[39mloads(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mtext, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n\u001b[1;32m    972\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m JSONDecodeError \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    973\u001b[0m     \u001b[38;5;66;03m# Catch JSON-related errors and raise as requests.JSONDecodeError\u001b[39;00m\n\u001b[1;32m    974\u001b[0m     \u001b[38;5;66;03m# This aliases json.JSONDecodeError and simplejson.JSONDecodeError\u001b[39;00m\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/json/__init__.py:346\u001b[0m, in \u001b[0;36mloads\u001b[0;34m(s, cls, object_hook, parse_float, parse_int, parse_constant, object_pairs_hook, **kw)\u001b[0m\n\u001b[1;32m    343\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m (\u001b[38;5;28mcls\u001b[39m \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m object_hook \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m\n\u001b[1;32m    344\u001b[0m         parse_int \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m parse_float \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m\n\u001b[1;32m    345\u001b[0m         parse_constant \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m object_pairs_hook \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m kw):\n\u001b[0;32m--> 346\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m _default_decoder\u001b[38;5;241m.\u001b[39mdecode(s)\n\u001b[1;32m    347\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mcls\u001b[39m \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/json/decoder.py:337\u001b[0m, in \u001b[0;36mJSONDecoder.decode\u001b[0;34m(self, s, _w)\u001b[0m\n\u001b[1;32m    333\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"Return the Python representation of ``s`` (a ``str`` instance\u001b[39;00m\n\u001b[1;32m    334\u001b[0m \u001b[38;5;124;03mcontaining a JSON document).\u001b[39;00m\n\u001b[1;32m    335\u001b[0m \n\u001b[1;32m    336\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[0;32m--> 337\u001b[0m obj, end \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mraw_decode(s, idx\u001b[38;5;241m=\u001b[39m_w(s, \u001b[38;5;241m0\u001b[39m)\u001b[38;5;241m.\u001b[39mend())\n\u001b[1;32m    338\u001b[0m end \u001b[38;5;241m=\u001b[39m _w(s, end)\u001b[38;5;241m.\u001b[39mend()\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/json/decoder.py:355\u001b[0m, in \u001b[0;36mJSONDecoder.raw_decode\u001b[0;34m(self, s, idx)\u001b[0m\n\u001b[1;32m    354\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mStopIteration\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n\u001b[0;32m--> 355\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m JSONDecodeError(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mExpecting value\u001b[39m\u001b[38;5;124m\"\u001b[39m, s, err\u001b[38;5;241m.\u001b[39mvalue) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m    356\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m obj, end\n",
      "\u001b[0;31mJSONDecodeError\u001b[0m: Expecting value: line 1 column 1 (char 0)",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mJSONDecodeError\u001b[0m                           Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[4], line 14\u001b[0m\n\u001b[1;32m      2\u001b[0m api_request_json \u001b[38;5;241m=\u001b[39m {\n\u001b[1;32m      3\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmodel\u001b[39m\u001b[38;5;124m\"\u001b[39m: _MODEL,\n\u001b[1;32m      4\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmessages\u001b[39m\u001b[38;5;124m\"\u001b[39m: [\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     10\u001b[0m     ],\n\u001b[1;32m     11\u001b[0m }\n\u001b[1;32m     13\u001b[0m \u001b[38;5;66;03m# Make your request and handle the response\u001b[39;00m\n\u001b[0;32m---> 14\u001b[0m response \u001b[38;5;241m=\u001b[39m llama\u001b[38;5;241m.\u001b[39mrun(api_request_json)\n\u001b[1;32m     15\u001b[0m resp \u001b[38;5;241m=\u001b[39m response\u001b[38;5;241m.\u001b[39mjson()\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/llamaapi/llamaapi.py:67\u001b[0m, in \u001b[0;36mLlamaAPI.run\u001b[0;34m(self, api_request_json)\u001b[0m\n\u001b[1;32m     65\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mrun_stream(api_request_json)\n\u001b[1;32m     66\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m---> 67\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mrun_sync(api_request_json)\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/llamaapi/llamaapi.py:53\u001b[0m, in \u001b[0;36mLlamaAPI.run_sync\u001b[0;34m(self, api_request_json)\u001b[0m\n\u001b[1;32m     51\u001b[0m response \u001b[38;5;241m=\u001b[39m requests\u001b[38;5;241m.\u001b[39mpost(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhostname\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdomain_path\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m, headers\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mheaders, json\u001b[38;5;241m=\u001b[39mapi_request_json)\n\u001b[1;32m     52\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m response\u001b[38;5;241m.\u001b[39mstatus_code \u001b[38;5;241m!=\u001b[39m \u001b[38;5;241m200\u001b[39m:        \n\u001b[0;32m---> 53\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mPOST \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mresponse\u001b[38;5;241m.\u001b[39mstatus_code\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mresponse\u001b[38;5;241m.\u001b[39mjson()[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mdetail\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     54\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m response\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/requests/models.py:975\u001b[0m, in \u001b[0;36mResponse.json\u001b[0;34m(self, **kwargs)\u001b[0m\n\u001b[1;32m    971\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m complexjson\u001b[38;5;241m.\u001b[39mloads(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mtext, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n\u001b[1;32m    972\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m JSONDecodeError \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    973\u001b[0m     \u001b[38;5;66;03m# Catch JSON-related errors and raise as requests.JSONDecodeError\u001b[39;00m\n\u001b[1;32m    974\u001b[0m     \u001b[38;5;66;03m# This aliases json.JSONDecodeError and simplejson.JSONDecodeError\u001b[39;00m\n\u001b[0;32m--> 975\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m RequestsJSONDecodeError(e\u001b[38;5;241m.\u001b[39mmsg, e\u001b[38;5;241m.\u001b[39mdoc, e\u001b[38;5;241m.\u001b[39mpos)\n",
      "\u001b[0;31mJSONDecodeError\u001b[0m: Expecting value: line 1 column 1 (char 0)"
     ]
    }
   ],
   "source": [
    "# API Request JSON Cell\n",
    "api_request_json = {\n",
    "    \"model\": _MODEL,\n",
    "    \"messages\": [\n",
    "        {\"role\": \"system\", \"content\": \"You are a serious assistant\"},\n",
    "        {\n",
    "            \"role\": \"user\",\n",
    "            \"content\": text,\n",
    "        },\n",
    "    ],\n",
    "}\n",
    "\n",
    "# Make your request and handle the response\n",
    "response = llama.run(api_request_json)\n",
    "resp = response.json()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\"Nel mezzo del cammin di nostri mortali giorni,\n",
      "mingendo con la mia arte, che alla misura\n",
      "del mio bel poco sia giunta, e son d'oggi il libro\" (Inferno, canto primo, versetti 1-3).\n"
     ]
    }
   ],
   "source": [
    "print(resp[\"choices\"][0][\"message\"][\"content\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
